### PR TITLE
Thumbnails as jpg instead of png to improve performance

### PR DESCRIPTION
**Pull request by diveflo**
As discussed in https://github.com/the-paperless-project/paperless/pull/559 this PR suggests to move the thumbnails from pngs to jpgs.

- file sizes are ~15-50% smaller
- jpg is widely adopted and compatible (in contrast to WebP as in #559 
- migration script to move over existing installs from png to jpg
- tested on Docker setup with and without enabled encryption

Let me know what you think!

### CODE REVIEW

File: src/documents/migrations/0023_thumbnails_to_jpeg.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        thumb_full = os.path.join(settings.MEDIA_ROOT,
+                                  "documents",
+                                  "thumbnails",
+                                  thumb)
+
+        if thumb_full.endswith(".jpg") or thumb_full.endswith(".jpg.gpg"):
+            continue
+        if thumb_full.endswith(".png"):
+            new_thumb = thumb_full[:-4] + ".jpg"
+            convert_png_to_jpg(thumb_full, new_thumb)
```
[the remark this answers is not in the document]
Comment (diveflo) [reply src/documents/migrations/0023_thumbnails_to_jpeg.py]: Aren’t the PNGs lossless? At least the optipng documentation states that it doesn’t lose any information. Or is it the original pdf -> png step? 
That one I haven’t checked to be honest - I was just going by the fact that many of my png thumbnails are actually bigger than the pdf itself. 

I’ll look into it - thanks for bringing this up!

[the remark this answers is not in the document]
Comment (danielquinn) [reply src/documents/migrations/0023_thumbnails_to_jpeg.py]: I'm far from an expert on image formats, but I *believe* that there's two ways for PNGs to store data and it depends on which one you're talking about when it comes to losslessness.  PNGs can be vector or raster (I think), and vectors are lossless, but I can't imagine raster images being lossless.

Either that, or PNG is only ever Vector, in which case making thumbnails out of them is a *terrible idea* and we should definitely be using GIF of JPEG.

**Comment by sbrunner** [comment]
@danielquinn Effectively I put a -1 because I suspect that we have a gain of size, but we lost in quality...
And I finally agree with all your previous comment @danielquinn :-)

File: src/documents/migrations/0023_thumbnails_to_jpeg.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sbrunner) [reply src/documents/migrations/0023_thumbnails_to_jpeg.py]: PNG is a raster-graphics file-format that supports lossless data compression (copy from Wikipedia), then he doesn't store any vector data.
He is a lossless image compression, but he can (and many times do) reduce the number of bits used to store a pixel (see: Pixel format: https://en.wikipedia.org/wiki/Portable_Network_Graphics#File_format)

**Comment by sbrunner** [comment]
Currently, we store the thumbnails as en image of 500px width, but we don't display at this size, shouldn’t we reduce the thumbnails images size? Or is it for hi DPI screens?

File: src/documents/migrations/0023_thumbnails_to_jpeg.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielquinn) [reply src/documents/migrations/0023_thumbnails_to_jpeg.py]: Neat!  Alright, so that sounds like this migration is just fine.  Ignore me :-)

**Comment by danielquinn** [comment]
I opted for 500px at the time because I thought there might be future use for this size.  Projects like [paperless-desktop](https://github.com/thomasbrueggemann/paperless-desktop) for example don't make use of the (admittedly limited & ugly) default UI and instead have a fancy electron-based interface.  In these cases, a larger thumbnail might make sense.  I'd also like to (one day) rewrite all of the UI in something like React that could make use of these thumbnails... but that's a "one day in the future" case that may never come.

The question I really want to answer out of all of this is whether or not this contributes a measurable improvement in performance and/or disk usage and whether the quality is something to be worried about.  To that end, I've taken page 2 out of my (lengthy) UK visa application process and used ImageMagick to covert it to a 500x500 thumbnail in both jpeg and png:

test (13k)
test (52k)

From my eye, they look about the same, and the resulting file size is roughly **4x** smaller on the jpeg side.  So for the disk space argument, I think we have a clear winner here.  Whether or not disk space really matters though is something I'd like another opinion on as I'm no longer the sole authority on this project.

As for performance, I'm not sure how to test for this outside of running this branch locally, running the migration, and then testing the "feel" of the interface as a result of this change.  I have a feeling that regardless of the file size, that the biggest performance bottleneck will be solved with the recent change to adopt Whitenoise and your other PR for gunicorn.  Again, I'm willing to be persuaded on this one.

Basically, I like what you've done here, the code changes are clean and easy to follow.  I'm concerned about the implications of a potentially long migration that'll change all the thumbnails on other people's systems though, and generally I'm hesitant to adopt Big Changes if the payoff isn't worth it, so I just wanna be sure before I offer my :+1:

**Comment by diveflo** [comment]
Sooo, I had a deeper look.
### Quality
I thought I compared the quality of the resulting jpgs between converting directly from pdf to jpg and png to jpg. But apparently I didn't o0 - so even if we go ahead with this, I will re-write the migration step to take the original pdf as input rather than the already converted png.

### Size
JPEG is definately smaller than png. The reduction is very dependent on the content. I just now ran an experiment with a part of the Python docs (~120 pages):

- original png: 14.9 MB
- png to jpg: 9.44 MB
- direct jpg: 7.2 MB

=> Another point in favor of going with the direct pdf to jpg conversion

If anyone wants to have a look, here's the resulting [thumbnails](https://file.io/asreNV).

### Performance
I started looking into the thumbnail data format because the Chrome DevTools Auditing feature suggested this as a major contributor to long loading times on paperless. However, my local setup runs on a RockPi4 which has just 2GB of RAM, so maybe this isn't such a pressing factor on less memory-constrained hardware.

As I see it, we have these options:

- stay with png
- stay with png but reduce thumbnail width to 180px (future UI permitting)
- move to jpg (after my re-write of the migration step)
- look into a cleaner (code) for providing a more modern format like webp (see #559)

The last three could also be activated via a switch in the configuration for people looking for improved performance, especially on constrained hardware.

Thanks for the thoughtful feedback and conversation - stuff like this is why I like open source:)

**Comment by sbrunner** [comment]
Thanks for your investigations, I remove my -1 :-)

**Comment by danielquinn** [comment]
Alright, of the 4 options @diveflo has presented, I'd opt for "stay with png" or "move to jpg", but I have no strong feelings around one or the other.  I'm going to request review from the other nerds on this project in the hopes that we can form some sort of consensus of what should happen.  Thank you for being so patient around this particular PR.  I just want to make sure we're making the right decision.

**Comment by MasterofJOKers** [comment]
I have to admit, that I have no strong feelings on this one. JPEG always had more artifacts for my use-cases, but I feel like it doesn't matter so much for thumbnails. So if it helps someone, why not.

Regarding the thumbnail sizes, we could also look into some of [the plugins available for django](https://code.djangoproject.com/wiki/ThumbNails) to generate thumbnails on the fly in the appropriate size. Maybe even expose that as an endpoint for third-party UIs. This would probably help with browser-performance, as the browser doesn't have to downscale all the images.

**Comment by diveflo** [comment]
> I have to admit, that I have no strong feelings on this one. JPEG always had more artifacts for my use-cases, but I feel like it doesn't matter so much for thumbnails. So if it helps someone, why not.
> 
> Regarding the thumbnail sizes, we could also look into some of [the plugins available for django](https://code.djangoproject.com/wiki/ThumbNails) to generate thumbnails on the fly in the appropriate size. Maybe even expose that as an endpoint for third-party UIs. This would probably help with browser-performance, as the browser doesn't have to downscale all the images.

That might actually be the cleaner solution. I'll close this pull request. Maybe I'll have some time over the holidays to look into this:)
